exercises/Daniel_Murphy/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,264 @@
+{
+  std::vector<int> type_vec;
+  int number = 1;
+  for (auto const& el : shapes)
+  {
+    std::string shape_type = el->get_type();
+    if (shape_type == type_requested)
+    {
+      type_vec.push_back(number);
+    }
+    number++;
+  }
+  if (type_vec.size() == 0)
+  {
+    std::cout << "None of shape type " << type_requested << " were found.";
+  }
+  else
+  {
+    std::cout << "The shapes of type " << type_requested << " are:  ";
+    for (auto const& el: type_vec)
+    {
+      std::cout << el << "  ";
+    }
+  }
+  std::cout << "\n";
+}
+
+void ShapeSorter::match_number_of_sides(int sides_requested)
+{
+  std::string plural = "sides";
+  if (sides_requested == 1)
+  {
+    plural = "side";
+  }
+  std::vector<int> side_vec;
+  int number = 1;
+  for (auto const& el: shapes)
+  {
+    int num_sides = el->get_sides();
+    if (num_sides == sides_requested)
+    {
+      side_vec.push_back(number);
+    }
+    number++;
+  }
+  if (side_vec.size() == 0)
+  {
+    std::cout << "No shapes with " << sides_requested << " sides were found.";
+  }
+  else
+  {
+    std::cout << "The shapes with " << sides_requested << " " << plural << " are:  ";
+    for (auto const& el: side_vec)
+    {
+      std::cout << el << "  ";
+    }
+  }
+  std::cout << "\n";
+}
+
+void ShapeSorter::area_descending()
+{
+  std::cout << "\nShape Number      Area / sq u \n";
+  std::cout << "############      ###########\n\n";
+
+  std::map<double, int> area_map;
+  int number = 1;
+  for (auto const&  el : shapes)
+  {
+    double shape_area = el->get_area();
+    area_map[shape_area] = number;
+    number++;
+  }
+
+  // Print areas in descending (reverse) order
+  for (auto it = std::next(area_map.end(),1); it != std::prev(area_map.begin(),1); --it)
+  {
+    int n = it->second;
+    double a = it->first;
+    size_t num_len = std::to_string(n).length();
+    size_t word_len = std::to_string(a).length();
+    std::cout << std::string(7-num_len, ' ') << n << std::string(25-word_len, ' ') << a << "\n";
+  }
+  std::cout << "\n";
+}
+
+void ShapeSorter::perimeter_descending()
+{
+  std::cout << "\nShape Number      Perimeter / u \n";
+  std::cout << "############      #############\n\n";
+
+  std::map<double, int> perimeter_map;
+  int number = 1;
+  for (auto const&  el : shapes)
+  {
+    double shape_perimeter = el->get_perimeter();
+    perimeter_map[shape_perimeter] = number;
+    number++;
+  }
+  // Print perimeters in descending (reverse) order
+  for (auto it = std::next(perimeter_map.end(),1); it != std::prev(perimeter_map.begin(),1); --it)
+  {
+    int n = it->second;
+    double a = it->first;
+    size_t num_len = std::to_string(n).length();
+    size_t word_len = std::to_string(a).length();
+    std::cout << std::string(7-num_len, ' ') << n << std::string(25-word_len, ' ') << a << "\n";
+  }
+  std::cout << "\n";
+}
+
+int main()
+{
+  ShapeSorter o;
+
+  Triangle a = {1.56,7.8}; o.add_new(a);
```

Comment:
I've made a comment against main.h on a v similar topic, but I don't think there's much point in creating the shape `a` here and then creating a second triangle inside add_new (which is what's happening when std::make_unique is called). You may as well just create it once inside std::make_unique and move that call into this main() method